core/arch/instrlist.c

```diff
@@ -481,15 +481,19 @@ instrlist_encode_to_copy(dcontext_t *dcontext, instrlist_t *ilist, byte *copy_pc
     instr_t *inst;
     int len = 0;
 #ifdef ARM
-    /* XXX i#1734: reset encdoe state to avoid any stale encode state
+    /* XXX i#1734: reset encode state to avoid any stale encode state
      * or dangling pointer.
      */
     if (instr_get_isa_mode(instrlist_first(ilist)) == DR_ISA_ARM_THUMB)
         encode_reset_it_block(dcontext);
 #endif
-    if (has_instr_jmp_targets || max_pc != NULL) {
+    if (IF_DEBUG_ELSE(true, has_instr_jmp_targets) || max_pc != NULL) {
         /* must set note fields first with offset, or compute length */
         for (inst = instrlist_first(ilist); inst; inst = instr_get_next(inst)) {
+            if (IF_DEBUG_ELSE(!has_instr_jmp_targets && instr_is_label(inst), false)) {
+                SYSLOG_INTERNAL_WARNING(
+                        "has_instr_jmp_targets was unset but a label was found");
+            }
```

Comment:
Hmm, this may end up being confusing: everything will work in debug (with a warning), but won't in release.  I was thinking we could detect most instances of this during actual encoding, though it's not foolproof: look for instr_t.note being zero in encode_operand (e.g., at x86/encode.c:1822).  Maybe *also* have a DOCHECK here that does a walk through the instrlist, but at a higher checklevel?